datafusion/sql/src/planner.rs

```diff
@@ -401,6 +401,14 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
     }
 
     pub(crate) fn convert_data_type(&self, sql_type: &SQLDataType) -> Result<DataType> {
+        // First check if any of the registered expr_planners can handle this type
+        for expr_planner in self.context_provider.get_expr_planners() {
```

Comment:
It seems strange that the planner for individual expressions would handle query-wide data types (though I realize this is currently the only hook that is available into the sql planner)

What would you think about adding a `DataTypePlanner` trait / field to the planner explicitly for this conversion?

Reply:
An independent trait looks good to me.  I'm thinking what's a good way to register the custom planner.
Maybe introduce a method in `ContextProvder::get_data_type_planner`?

```rust
pub trait ContextProvider {
...
  fn get_data_type(&self) -> Option<Arc<dyn DataTypePlanner>>
      None
  }
}
```

We can allow registering the data type planner in `SessionState`. WDYT?